### PR TITLE
feat(anomaly): add normalization methods for anomaly detection scores

### PR DESCRIPTION
**Pull request by rgthelen**
## Summary

Add three score normalization methods for anomaly detection, allowing users to choose how raw anomaly scores are transformed for easier interpretation and threshold setting.

### Normalization Methods

| Method | Score Range | Default Threshold | Best For |
|--------|-------------|-------------------|----------|
| `standardization` (default) | 0-1 | 0.5 | General use, bounded scores |
| `zscore` | Unbounded (std devs) | 2.0 | Statistical interpretation |
| `raw` | Backend-specific | 0.0 (set your own) | Debugging, advanced users |

### Key Changes

- **AnomalyModel**: Added `normalization` parameter and `NormalizationMethod` type
- **API Endpoints**: Added `normalization` parameter to `/v1/anomaly/fit`, `/v1/anomaly/score`, `/v1/anomaly/detect`
- **Persistence**: Normalization method and statistics are saved/loaded with models
- **Documentation**: Comprehensive guide with use cases, algorithm comparison, and troubleshooting

### Example Usage

```bash
# Z-score: flag anything > 3 standard deviations
curl -X POST http://localhost:11540/v1/anomaly/fit \
  -d '{"model": "sensor", "normalization": "zscore", "data": [...], "threshold": 3.0}'

# Result: scores like "3.25 std devs" are immediately interpretable
```

### Why This Matters

- **Standardization**: Bounded 0-1 scores are easy to understand and compare
- **Z-Score**: Engineers can say "this is 3 standard deviations from normal" 
- **Raw**: Full control for debugging or replicating research

## Test Plan

- [x] Run `./examples/ocr_and_document/test_anomaly.sh` - tests all 3 methods
- [x] Verify all backends work with mixed text data
- [x] Verify model save/load preserves normalization settings
- [ ] Manual testing of API endpoints

___

### **PR Type**
Enhancement

___

### **Description**
- Add three score normalization methods (standardization, zscore, raw) for anomaly detection

- Persist normalization settings and statistics with saved models

- Add normalization parameter to all anomaly API endpoints

- Comprehensive documentation with use cases and algorithm comparisons

- Update test script to demonstrate all three normalization methods

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Raw Anomaly Scores"] --> B["Normalization Method"]
  B --> C1["Standardization<br/>0-1 range"]
  B --> C2["Z-Score<br/>Std Deviations"]
  B --> C3["Raw<br/>Backend Native"]
  C1 --> D["Normalized Scores<br/>+ Threshold"]
  C2 --> D
  C3 --> D
  D --> E["Anomaly Detection<br/>Results"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>anomaly_model.py</strong><dd><code>Implement three score normalization methods with persistence</code></dd></summary>
<hr>

runtimes/universal/models/anomaly_model.py

<ul><li>Add <code>NormalizationMethod</code> type with three options: standardization, <br>zscore, raw<br> <li> Add <code>normalization</code> parameter to <code>AnomalyModel.__init__()</code> with default <br>"standardization"<br> <li> Store normalization statistics during fit: median/IQR for <br>standardization, mean/std for zscore<br> <li> Implement <code>_normalize_scores()</code> method supporting all three <br>normalization approaches<br> <li> Persist normalization method and statistics in model save/load for <br>both PyTorch and sklearn backends<br> <li> Update <code>threshold</code> property to return method-specific defaults (0.5 for <br>standardization, 2.0 for zscore, 0.0 for raw)<br> <li> Add comprehensive docstrings explaining each normalization method with <br>examples</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/624/files#diff-415c7bd8c2b5eea32d3fac8fad30d787b4905853b388c85d6df5942f130a9140">+185/-29</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>server.py</strong><dd><code>Add normalization parameter to API endpoints</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

runtimes/universal/server.py

<ul><li>Add <code>normalization</code> parameter to <code>load_anomaly()</code> function<br> <li> Add <code>normalization</code> field to <code>AnomalyScoreRequest</code> and <code>AnomalyFitRequest</code> <br>Pydantic models<br> <li> Pass normalization parameter through all anomaly endpoints: <br><code>/v1/anomaly/fit</code>, <code>/v1/anomaly/score</code>, <code>/v1/anomaly/detect</code><br> <li> Add documentation comments explaining normalization methods in request <br>classes<br> <li> Update endpoint docstrings with normalization parameter details</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/624/files#diff-eb7e358ca4b36ecd277661e724c825dfbf087b8bbb6aa4accf1a9eb72cc1590f">+22/-1</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_anomaly.sh</strong><dd><code>Refactor test script to demonstrate normalization methods</code></dd></summary>
<hr>

examples/ocr_and_document/test_anomaly.sh

<ul><li>Restructure test script to demonstrate all three normalization methods <br>instead of production workflow<br> <li> Add Test 1: Standardization method with 0-1 range scoring<br> <li> Add Test 2: Z-Score method with standard deviation interpretation<br> <li> Add Test 3: Raw score method with backend-native scores<br> <li> Add Test 4: Side-by-side comparison of all three methods on same data<br> <li> Add Test 5: Compare different backends (Isolation Forest, One Class <br>SVM, LOF) with zscore normalization<br> <li> Add Test 6: Save and load model workflow<br> <li> Improve output formatting with color-coded results and interpretable <br>summaries<br> <li> Replace CSV-based tests with simpler temperature sensor data examples</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/624/files#diff-918aa02f632d5da0ef5b038cc3a1e0a3cc5338cc52203921c323dc401f04b691">+401/-377</a></td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>anomaly-detection.md</strong><dd><code>Comprehensive documentation for normalization methods</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/website/docs/models/anomaly-detection.md

<ul><li>Add normalization methods to feature list in introduction<br> <li> Add comprehensive "Score Normalization Methods" section with <br>comparison table<br> <li> Document standardization method: sigmoid transformation, 0-1 range, <br>threshold guidance<br> <li> Document zscore method: standard deviation interpretation, threshold <br>guidance (2.0=unusual, 3.0=rare, 4.0+=extreme)<br> <li> Document raw method: backend-native scores with per-backend range <br>table<br> <li> Add decision matrix for choosing normalization method by use case<br> <li> Expand algorithm documentation with detailed explanations, <br>strengths/limitations, training requirements<br> <li> Add algorithm comparison table (speed, dimensions, sample <br>requirements, etc.)<br> <li> Add use case examples: API log detection, sensor monitoring, financial <br>fraud, network intrusion<br> <li> Add troubleshooting section for common normalization issues<br> <li> Update API examples to include normalization parameter<br> <li> Update best practices with normalization-aware threshold tuning</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/624/files#diff-8390ff378c9147521cc5e6811bd31783195347ca41956625016fa04aab163726">+391/-49</a></td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by rgthelen** [comment]
This PR addresses:
- #641 - Anomaly detection scores are difficult to interpret and set thresholds for
- #640 - Normalization parameter was silently ignored when model cached with different method

**Comment by rgthelen** [comment]
@claude review this PR

File: designer/src/types/ml.ts [hunk 1]
```diff
@@ -111,6 +114,64 @@ export const ANOMALY_BACKEND_DISPLAY: Record<AnomalyBackend, string> = {
...
+  description: string
+  defaultThreshold: number
+}
+
+export const NORMALIZATION_OPTIONS: NormalizationOption[] = [
+  {
+    value: 'standardization',
+    label: 'Standardized (0-1)',
+    description: 'Scores 0-1, threshold ~0.6. Best for general use.',
+    defaultThreshold: 0.6,
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply designer/src/types/ml.ts:151]: nope. this is fine.

File: designer/src/components/Models/TrainedModels.tsx [hunk 2]
```diff
@@ -30,10 +30,11 @@ function toTrainedModel(
   type: TrainedModelType,
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply designer/src/components/Models/TrainedModels.tsx:30]: no, it works as is.

**Comment by rgthelen** [comment]
Like claude suggested, I created an issue to add the ability to override the normalization method at inference time.
